### ISSUE TITLE
false positives for Style/HashEachMethods

### ISSUE THREAD

**Issue by dinoex** [rubocop/rubocop#13678]
rubocop reports  a false positive for Style/HashEachMethods

Using autocorrect the generated code does raise "undefined method".

--------

## Expected behavior

no warning should be reported with the array.

Autocorrect should not generate broken code.

```
# ruby -w bug-each.rb
unused: a1
unused: b1
```     

## Actual behavior

robocop issues a warning.

```
# rubocop -a --debug bin/bug-each.rb
For /media/xxxx: Default configuration from /home/xxx/.gem/ruby/3.2/gems/rubocop-1.70.0/config/default.yml
Use parallel by default.
Skipping parallel inspection: only a single file needs inspection
Inspecting 1 file
Scanning /media/xxx/bin/bug-each.rb
Loading cache from /home/xxx/.cache/rubocop_cache/47eb1bac1730574d2f716a7237ce947acab28e62/6d7a3b621ca1730e04accd938619e4bdab66cfb1/e38131b8756a8e600cf2267f48e01c82f6385e9f
C

Offenses:

bin/bug-each.rb:28:3: C: [Correctable] Style/HashEachMethods: Use each_key instead of each and remove the unused _dir block argument.
  @subs.each do |name, _dir| ...
  ^^^^^^^^^^^^^^^^^^^^^^^^^^

1 file inspected, 1 offense detected, 1 more offense can be corrected with `rubocop -A`
Finished in 0.3739329678937793 seconds
```

Running the corrected script fails:

```
# ruby -w bug-each.rb
bug-each.rb:27:in `report_unused_fixes': undefined method `each_key' for [["a1", "a2"]]:Array (NoMethodError)

  @subs.each_key do |name|
       ^^^^^^^^^
Did you mean?  each_entry
        from bin/bug-each.rb:42:in `<main>'
```

## Steps to reproduce the problem

run rubocop with this file as "bug-each.rb"

```rb
#!/usr/local/bin/ruby

def read_fixes
  @seen_subs = {}
  @seen_fixes = {}
  @fixes = { 'b1' => 'b2' }
  @subs = [ [ 'a1', 'a2' ] ]
end

def clean_key( key )
  @subs.each do |name, dir|
    next unless key.include?( name )

    @seen_subs[ name ] = true
    key.sub!( name, dir )
  end

  if @fixes.key?( key )
    @seen_fixes[ key ] = true
    key = @fixes[ key ].dup
  end

  key
end

def report_unused_fixes
  @subs.each do |name, _dir|
    next if @seen_subs.key?( name )

    puts "unused: #{name}"
  end

  @fixes.each_key do |name|
    next if @seen_fixes.key?( name )

    puts "unused: #{name}"
  end
end

read_fixes
clean_key( 'test' )
report_unused_fixes

# eof

```

## RuboCop version

```
# rubocop -V
1.70.0 (using Parser 3.3.5.0, rubocop-ast 1.37.0, analyzing as Ruby 2.7, running on ruby 3.2.5) [amd64-freebsd14]
```
Same issue happens also with rubocop 1.67.0.

**Comment by Earlopain** [comment]
Hi, this is expected and documented in the cops safety section. You can either disable the offense with an inline comment or write it like this:

```rb
@subs.each do |(name, _dir)|
end
```

to make it clear that it is an array of arrays instead of an hash you are iterating over.

**Comment by dvandersluis** [comment]
In addition to what @Earlopain, be aware that `-A` performs *all* autocorrects, regardless of safety. This can result in incorrect autocorrects depending on context (eg. rubocop cannot know that `@subs` in your original code doesn't respond to `each_key`). Note that your original rubocop output does not correct this code but marks it as "Correctable". Using `rubocop -a` will only run autocorrections that are safe.
